metaseq/data/document_to_sequence.py

```diff
@@ -0,0 +1,352 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+from typing import Optional
+
+import numpy as np
+import torch
+
+from metaseq.data import data_utils
+import time
+from metaseq.data.atomic_array import AtomicArray
+
+
+# gotta be a better way to get this...
+def f():
+    x = 4
+    return lambda: x
+
+
+Cell = type(f().__closure__[0])
+
+
+class DocumentToSequenceDataset(torch.utils.data.IterableDataset):
+    """Take a dataset containing documents and turn it into an iterable dataset
+    returning sequences of block_size tokens.
+
+    This dataset can only be iterated over once.
+
+    Documents are optionally permuted (permute_documents).
+    This iterator has a `len_cache` which is an AtomicArray mapping
+    document id to the number of tokens in the document, which it populates.
+    When populated the len_cache allows the iterator to quickly skip the first to_skip sequences.
+
+    Args:
+        dataset (~torch.utils.data.IterableDataset): dataset to chunk
+        block_size (int): maximum block size
+        break_mode (str, optional): Mode used for breaking tokens. Values can
+            be one of:
+            - 'none': break tokens into equally sized blocks (up to block_size)
+        drop_last (bool, optional): drop the last item (default: False)
+        padding_idx (int, optional): index to use for padding symbols
+            (required if *drop_last* is ``False``)
+        shuffle_buffer_size (int, optional): buffer this many items and shuffle
+            using the provided *seed*; default value is 1, so no shuffling is
+            performed. This can be adjusted dynamically after initialization,
+            but only before iteration has begun.
+        seed (int, optional): seed for shuffling
+        permute_documents (bool, optional): randomly permute the order the documents are read (default: True)
+        to_skip (int, optional): skip the first to_skip sequences (Default: 0)
+    """
+
+    def __init__(
+        self,
+        dataset: torch.utils.data.IterableDataset,
+        block_size: int,
+        break_mode: str = "none",
+        drop_last: Optional[bool] = False,
+        padding_idx: Optional[int] = None,
+        shuffle_buffer_size: int = 1,
+        seed: Optional[int] = None,
+        len_cache=None,
+        to_skip=0,
+        permute_documents=True,
+    ):
+        super().__init__()
+        self.dataset = dataset
+        assert len(dataset) > 0
+        self.worker_offset = 0
```

Comment:
Yeah, I will add a better comment here. What is happening:
* PyTorches data loader object pulls data round-robin from each worker, but it always starts with the _first_ worker when created, without any way to configure it.
* When we restart, we might have to restart by pulling data from a non-first worker (if n is not divisible by 8). Since we can't change how the PyTorch data loader object pulls data from the workers, we change the workers instead so that the first worker is the one we want to pull data from next.